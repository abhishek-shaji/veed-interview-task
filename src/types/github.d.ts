export interface GithubRepositoryDetails {
  id: number;
  name: string;
  html_url: string;
  full_name: string;
  description: string;
  forks_count: number;
  stargazers_count: number;
  watchers_count: number;
}

export interface GithubApiResponse {
  total_count: number;
  incomplete_results: boolean;
  items: GithubRepositoryDetails[];
}

export interface FormattedRepositoryDetails {
  id: number;
  name: string;
  description: string;
  full_name: string;
  url: string;
  stars: number;
  watchers: number;
  forks: number;
}
