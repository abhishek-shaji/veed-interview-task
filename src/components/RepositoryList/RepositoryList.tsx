import React from 'react';

import { FormattedRepositoryDetails } from '@/types/github';
import { DatabaseIcon } from '@heroicons/react/outline';
import { StarIcon } from '@heroicons/react/solid';
import { LoadingSpinner } from '@/components/Icons/LoadingSpinner';

interface CompanyListPropType {
  items: FormattedRepositoryDetails[];
  isLoading?: boolean;
  toggleFavourite: (id: string) => any;
  favourites: string[];
}

const RepositoryList = ({
  items,
  isLoading,
  favourites,
  toggleFavourite,
}: CompanyListPropType) => {
  if (isLoading) {
    return (
      <div className="container flex justify-center py-12 mx-auto">
        <LoadingSpinner />
      </div>
    );
  }

  return (
    <div className="container mx-auto">
      <div className="flex justify-between my-5 py-3 mb-10 border-b border-slate-200">
        <h2 className="text-lg font-semibold">{items.length} Repositories</h2>
      </div>
      {items.map(
        ({
          id,
          name,
          full_name,
          url,
          description,
          stars,
          forks,
        }: FormattedRepositoryDetails) => (
          <a
            href={url}
            className="block border-b border-slate-200 hover:-translate-y-1 hover:shadow-lg cursor-pointer transition-all"
            key={id}
          >
            <div className="flex justify-between items-center">
              <div className="flex items-center my-5">
                <DatabaseIcon
                  className="w-20 h-20 rounded object-cover text-slate-600"
                  strokeWidth={1}
                />
                <div className="ml-2">
                  <div className="font-bold capitalize">{name}</div>
                  <div className="opacity-75 text-sm capitalize mb-1">
                    {description || 'No description'}
                  </div>
                  <div className="opacity-75 text-xs capitalize">
                    Stars: {stars}
                  </div>
                  <div className="opacity-75 text-xs capitalize">
                    Forks: {forks}
                  </div>
                </div>
              </div>
              <button
                className="text-sm text-slate-500 mr-6"
                onClick={(e) => {
                  e.preventDefault();

                  toggleFavourite(full_name);
                }}
              >
                <StarIcon
                  className={`h-8 w-8 rounded object-cover ${
                    favourites.includes(full_name)
                      ? 'text-amber-500 hover:text-amber-600'
                      : 'text-slate-300 hover:text-slate-600'
                  }`}
                />
              </button>
            </div>
          </a>
        ),
      )}
    </div>
  );
};

export { RepositoryList };
