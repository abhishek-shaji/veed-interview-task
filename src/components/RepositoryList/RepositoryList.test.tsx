import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { RepositoryList } from '@/components/RepositoryList/RepositoryList';
import { FormattedRepositoryDetails } from '@/types/github';

const mockRepository: FormattedRepositoryDetails = {
  id: 1,
  name: 'test',
  full_name: 'test',
  url: 'test',
  description: 'test',
  stars: 1,
  forks: 1,
  watchers: 1,
};

describe('RepositoryList', () => {
  it('should render correctly', () => {
    const { container } = render(
      <RepositoryList
        items={[]}
        favourites={[]}
        toggleFavourite={() => {}}
        isLoading={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loading', () => {
    const { container } = render(
      <RepositoryList
        items={[]}
        favourites={[]}
        toggleFavourite={() => {}}
        isLoading
      />,
    );

    const element = screen.getByTestId('loading-spinner');

    expect(container).toContainElement(element);

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with repositories', () => {
    const { container } = render(
      <RepositoryList
        items={[mockRepository]}
        favourites={[]}
        toggleFavourite={() => {}}
        isLoading={false}
      />,
    );

    const element = screen.getByTestId('repository-list-item');

    expect(container).toContainElement(element);

    expect(container).toMatchSnapshot();
  });
});
