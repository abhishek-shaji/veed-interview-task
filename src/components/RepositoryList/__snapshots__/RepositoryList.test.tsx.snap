// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryList should render correctly 1`] = `
<div>
  <div
    class="container mx-auto"
  >
    <div
      class="flex justify-between my-5 py-3 mb-10 border-b border-slate-200"
    >
      <h2
        class="text-lg font-semibold"
      >
        0
         Repositories
      </h2>
    </div>
  </div>
</div>
`;

exports[`RepositoryList should render correctly with loading 1`] = `
<div>
  <div
    class="container flex justify-center py-12 mx-auto"
  >
    <svg
      class="w-8 h-8 mr-2 text-slate-200 animate-spin fill-blue-600"
      data-testid="loading-spinner"
      fill="none"
      role="status"
      viewBox="0 0 100 101"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
        fill="currentColor"
      />
      <path
        d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
        fill="currentFill"
      />
    </svg>
  </div>
</div>
`;

exports[`RepositoryList should render correctly with repositories 1`] = `
<div>
  <div
    class="container mx-auto"
  >
    <div
      class="flex justify-between my-5 py-3 mb-10 border-b border-slate-200"
    >
      <h2
        class="text-lg font-semibold"
      >
        1
         Repositories
      </h2>
    </div>
    <a
      class="block border-b border-slate-200 hover:-translate-y-1 hover:shadow-lg cursor-pointer transition-all"
      data-testid="repository-list-item"
      href="test"
    >
      <div
        class="flex justify-between items-center"
      >
        <div
          class="flex items-center my-5"
        >
          <svg
            aria-hidden="true"
            class="w-20 h-20 rounded object-cover text-slate-600"
            fill="none"
            stroke="currentColor"
            stroke-width="1"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 7v10c0 2.21 3.582 4 8 4s8-1.79 8-4V7M4 7c0 2.21 3.582 4 8 4s8-1.79 8-4M4 7c0-2.21 3.582-4 8-4s8 1.79 8 4m0 5c0 2.21-3.582 4-8 4s-8-1.79-8-4"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
          <div
            class="ml-2"
          >
            <div
              class="font-bold capitalize"
            >
              test
            </div>
            <div
              class="opacity-75 text-sm capitalize mb-1"
            >
              test
            </div>
            <div
              class="opacity-75 text-xs capitalize"
            >
              Stars: 
              1
            </div>
            <div
              class="opacity-75 text-xs capitalize"
            >
              Forks: 
              1
            </div>
          </div>
        </div>
        <button
          class="text-sm text-slate-500 mr-6"
        >
          <svg
            aria-hidden="true"
            class="h-8 w-8 rounded object-cover text-slate-300 hover:text-slate-600"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
            />
          </svg>
        </button>
      </div>
    </a>
  </div>
</div>
`;
